### PR TITLE
PVF: Move PVF workers into separate crate

### PR DESCRIPTION
**Pull request by mrcnski**
# PULL REQUEST

## Overview

Moves the PVF workers into their own crate. This was motivated by https://github.com/paritytech/polkadot/issues/4718 and trying to move unnecessary dependencies out of the workers so we can sandbox the workers better. In particular there is a lot of IO stuff that may be coming from `tokio`'s `fs` feature. So I am extracting out the worker into a separate crate and building `tokio` without the `fs` feature.

Aside from that motivation, this is good to have anyway. It makes the host/worker separation much more clear.

## Companion

Cumulus companion: https://github.com/paritytech/cumulus/pull/2470

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2701590

File: Cargo.toml [hunk 1]
```diff
@@ -80,6 +80,7 @@ members = [
 	"node/core/parachains-inherent",
 	"node/core/provisioner",
 	"node/core/pvf",
+    "node/core/pvf/worker",
```
Comment (s0me0ne-unkn0wn) [inline Cargo.toml:83]: Nit: looks like indentation went wrong?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2704024

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2704197

**Comment by slumber** [comment]
Reading through the linked issue, if I understand correctly, 
> Build the worker binary in build.rs against musl.

means we want to run worker from a separate binary.
This is actually not easy because it makes polkadot node hard to ship.

This PR does a simple refactoring with extracting worker-related code into a separate crate, but won't help with the above since it still uses polkadot binary (cli is not updated thus currently it won't compile).
Was it the intention?

**Comment by s0me0ne-unkn0wn** [comment]
Good job! It took me some time to understand that it's not about extracting what we were calling "worker" into a separate crate, it's about splitting logic into two parts: the host part interacting with the worker and the worker itself. So the host part is in `node/core/pvf/src/execute/worker.rs`, and the worker part is in `node/code/pvf/worker/src/execute.rs`, and I feel my mind is already blowing while I'm writing this :exploding_head: I believe the former should be renamed into something like `worker_interface` or whatever, I'm not sure, naming is hard, but it should clearly state that it's not the worker itself but the host interface to the worker. Otherwise, the next developer exploring that code would want to send a curse upon us :grin:

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2704387

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2707023

File: node/core/pvf/worker/src/prepare.rs [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+///	2. Start a memory tracker in a separate thread.
+///
+///	3. Start the CPU time monitor loop and the actual preparation in two separate threads.
+///
+///	4. Select on the two threads created in step 3. If the CPU timeout was hit, the CPU time monitor
+///	   thread will trigger first.
+///
+///	5. Stop the memory tracker and get the stats.
+///
+/// 6. If compilation succeeded, write the compiled artifact into a temporary file.
```
Comment (slumber) [inline node/core/pvf/worker/src/prepare.rs:77]: identation (comes from the older code but anyway)

File: node/test/performance-test/Cargo.toml [hunk 3]
```diff
@@ -10,11 +10,14 @@ quote = "1.0.26"
...
 log = "0.4"
 
-polkadot-node-core-pvf = { path = "../../core/pvf" }
+polkadot-node-core-pvf-worker = { path = "../../core/pvf/worker" }
 polkadot-erasure-coding = { path = "../../../erasure-coding" }
 polkadot-node-primitives = { path = "../../primitives" }
 polkadot-primitives = { path = "../../../primitives" }
 
+sc-executor-common = { git = "https://github.com/paritytech/substrate", branch = "master" }
+sp-maybe-compressed-blob = { git = "https://github.com/paritytech/substrate", branch = "master" }
```
Comment (slumber) [inline node/test/performance-test/Cargo.toml:19]: Makes sense for the worker crate to `pub use` these two.

Comment (mrcnski) [reply node/test/performance-test/Cargo.toml:19]: Can you explain the reasoning? Seems to me like unnecessary indirection.

Comment (slumber) [reply node/test/performance-test/Cargo.toml:19]: Well, if every crate that uses `worker` also needs these 2, we might as well just re-export them. A matter of taste so feel free to disregard.

**Review by slumber — APPROVED** [review]

**Review by s0me0ne-unkn0wn — APPROVED** [review]
LVGTM, approving. Despite the discussion about worker separation still ongoing, and we're still trying to figure out if it's worthwhile, this PR is not solely intended to prepare the code for that process. It's about separating highly integrated logic, which took a lot of work to understand. I personally recall a struggle with that code where it was not possible to tell at first glance where the host part was and where the worker itself. So I believe this change is beneficial whether we implement worker separation in the end or not.

**Comment by mrcnski** [comment]
> I personally recall a struggle with that code where it was not possible to tell at first glance where the host part was and where the worker itself.

Me too! This should make things more clear for first-time readers.

**Comment by mrcnski** [comment]
bot merge

**Pull request merged**